File: frontend/src/shared/lib/ratings.ts
/**
 * Escala de notas, a mesma do backend (app/shared/ratings.py): de 0 a 10, e cada ponto vale
 * meia estrela (1 = ½ estrela, 10 = 5 estrelas).
 */

export const MAX_SCORE = 10
export const MAX_STARS = 5

/** Nota 0–10 → estrelas 0–5, arredondando para a meia estrela mais próxima (.5 sobe). */
export function scoreToStars(nota: number): number {
  return Math.floor(nota + 0.5) / 2
}

/** 4.5 → "4,5"; 4 → "4". */
export function formatStars(stars: number): string {
  return String(stars).replace('.', ',')
}

/** 4.5 → "4,5 estrelas"; 1.5 → "1,5 estrela" (em português, só a partir de 2 é plural). */
export function starsLabel(stars: number): string {
  return `${formatStars(stars)} ${stars > 0 && stars < 2 ? 'estrela' : 'estrelas'}`
}

File: backend/app/shared/ratings.py
"""Escala de notas usada por todas as avaliações.

As notas vão de 0 a 10 e cada ponto vale meia estrela: 1 = ½ estrela, 10 = 5 estrelas.
Novas avaliações recebem notas inteiras; as importadas do CSV podem ser decimais.
"""

import math
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, computed_field

MIN_SCORE = 0
MAX_SCORE = 10

Score = Annotated[int, Field(ge=MIN_SCORE, le=MAX_SCORE, strict=True)]


def score_to_stars(nota: float) -> float:
    """Converte uma nota 0–10 em estrelas 0–5, arredondando para a meia estrela mais próxima."""

    return math.floor(nota + 0.5) / 2


class RatingSummary(BaseModel):
    """Quantidade e média das avaliações recebidas, com a média também em estrelas."""

    model_config = ConfigDict(from_attributes=True)

    qtd_avaliacoes_usuarios: int = 0
    nota_media_usuarios: float | None = None

    @computed_field
    @property
    def estrelas_media(self) -> float | None:
        if self.nota_media_usuarios is None:
            return None
        return score_to_stars(self.nota_media_usuarios)
